### PR TITLE
Fix FileSet (and FileSet.Entry) Equality and Hash Implementations

### PR DESCRIPTION
**Pull request by michaelahlers**
Resolves issue #43.

### CODE REVIEW

**Comment by michaelahlers** [comment]
I can't explain a [buffer overflow that occurred](https://travis-ci.org/flapdoodle-oss/de.flapdoodle.embed.process/jobs/152516521#L1100) on [build 48](https://travis-ci.org/flapdoodle-oss/de.flapdoodle.embed.process/builds/152516520).

File: src/main/java/de/flapdoodle/embed/process/config/store/FileSet.java [hunk 1]
```diff
@@ -83,50 +83,83 @@ public boolean equals(Object other) {
...
+			return false;
+		}
+
+		if (!(other instanceof FileSet)) {
+			return false;
+		}
+
+		FileSet files = (FileSet) other;
+
+		return !(_entries != null ? !_entries.equals(files._entries) : files._entries != null);
```
Comment (matlockx) [inline src/main/java/de/flapdoodle/embed/process/config/store/FileSet.java]: can you please flip the operators? negate - negate is confusing

File: src/test/java/de/flapdoodle/embed/process/store/FileSetEntryTest.java [hunk 2]
```diff
@@ -0,0 +1,21 @@
+package de.flapdoodle.embed.process.store;
+
+import de.flapdoodle.embed.process.config.store.FileSet;
+import nl.jqno.equalsverifier.EqualsVerifier;
+import nl.jqno.equalsverifier.Warning;
+import org.junit.Test;
+
+/**
+ * @author [[mailto:michael@ahlers.consulting Michael Ahlers]]
```
Comment (matlockx) [inline src/test/java/de/flapdoodle/embed/process/store/FileSetEntryTest.java]: please remove header

File: src/main/java/de/flapdoodle/embed/process/config/store/FileSet.java [hunk 3]
```diff
@@ -83,50 +83,83 @@ public boolean equals(Object other) {
...
 				return false;
 			}
 
 			Entry entry = (Entry) other;
 
+			/* Because Pattern doesn't implement equals, use its underlying expression (String) for equality. See #43. */
 			return _type == entry._type
 				&& !(_destination != null ? !_destination.equals(entry._destination) : entry._destination != null)
-				&& !(_matchingPattern != null ? !_matchingPattern.equals(entry._matchingPattern) : entry._matchingPattern != null);
+				&& !(_matchingPattern != null ? !_matchingPattern.pattern().equals(null == entry._matchingPattern ? null : entry._matchingPattern.pattern()) : entry._matchingPattern != null);
```
Comment (matlockx) [inline src/main/java/de/flapdoodle/embed/process/config/store/FileSet.java]: i know the former line was also like that but could you please flip operators here?

**Comment by matlockx** [comment]
hm, the buffer overflow looks weird. @michaelmosmann any idea?

**Comment by michaelahlers** [comment]
@matlockx, sorry this slipped through the cracks (been swamped). I can create a new PR with those requested changes, if you like.

**Comment by michaelmosmann** [comment]
@michaelahlers hold on.. i did a lot of refactoring using immutables.org .. so we will get rid of a lot of handwritten code ...
but thanks for the hint, that Pattern does not implement hash & equal .. (wtf) .. so i could fix this:)

**Comment by michaelahlers** [comment]
> refactoring using immutables.org

Yes! Excited to see this. That'll cut down a ton of rote work in dependant libraries, too. (So tired of writing builders.)

> that Pattern does not implement hash & equal

It's totally surprising, especially because `Pattern` is documented as being immutable (and, therefore, thread safe). In other words, no reason _not_ to implement those methods.

**Comment by michaelmosmann** [comment]
@michaelahlers 

> that Pattern does not implement hash & equal

this does not surprise me anymore ... there should be some kind of warning sign .. :(
best wtf this year: oracle timestamp implements equal but NO hashcode ..

**Pull request merged**